### ISSUE TITLE
Better testing of warnings

### ISSUE THREAD

**Issue by momchil-flex** [flexcompute/tidy3d#1286]
Currently, in many tests we use [this function](https://github.com/flexcompute/tidy3d/blob/f60bf47f84264bc622e548fa41dfc344fb85ceaf/tests/utils.py#L610) to check if a warning was issued. However, I've encountered in the past tests that were passing because *a completely unrelated* warning was issued by the setup, while the target warning was not. This recently tripped up @shash-sharma too when writing a test. I feel like it is generally making our tests for warnings not very reliable.

Is there a way to test more explicitly that the right warning was raised? And if not, maybe at least we could test that exactly one warning was raised, or even provide a number of warnings that are expected as an argument to the function. I think we have some test where we check for a specific number of warnings?

**Comment by tylerflex** [comment]
How about an optional string argument to assert log level. If present we assert that the string is in the warning message?

**Comment by tylerflex** [comment]
I looked at the tests, see my comment https://github.com/flexcompute/tidy3d/pull/1288/files

Basically if you have code blocks and want to check if a **new** warning is thrown, you must do `log_capture.clear()` between them. Otherwise, the first warning will always trigger the second `assert_log_level`.

```py
# something that warns
assert_log_level(log_capture, "WARNING")
log_capture.clear()

# something else that warns
assert_log_level(log_capture, "WARNING")
log_capture.clear()
```

Of course this doesn't solve the issue about triggering specific warnings. I think a simple string matching approach should work fine. 

Another approach is to include in your test a "base" case that doesn't warn. And assert log level is None. See my tests for the nonlinear adjoint medium.

https://github.com/flexcompute/tidy3d/blob/9e152120a492390c194158837f4942b2265ea724/tests/test_plugins/test_adjoint.py#L1684-L1702

```py
    def test_log_level(desired_level):
        """Convenience function to test the log level and clear it."""
        assert_log_level(log_capture, desired_level)
        log_capture.clear()

    # no nonlinearity (no warning)
    test_log_level(None)

    # nonlinear simulation.medium (error)
    sim = sim_base.updated_copy(medium=nl_medium)
    test_log_level("WARNING")

    # nonlinear structure (warn)
    sim = sim_base.updated_copy(structures=[struct_static_nl])
    test_log_level("WARNING")

    # nonlinear input_structure (warn)
    sim = sim_base.updated_copy(input_structures=[input_struct_nl])
    test_log_level("WARNING")
```

**Comment by momchil-flex** [comment]
I think it's still best to do something like string matching or at least counting the number of warnings raised. Even the base test thing is prone to unexpected behavior.

**Comment by tylerflex** [comment]
We'd still need to clear the log capture each block of code i think. Otherwise you can't be sure that multiple warnings weren't triggered by a single block. Basically i think we need log clearing _and_ string matching for more assurance.

**Comment by momchil-flex** [comment]
Yeah agreed. It would be nice if we could make this a bit more foolproof too, in case a dev forgets to do it, but I'm not sure if there's a good way.

**Comment by tylerflex** [comment]
Context manager maybe?

```
with log_capture.assert_raises("warning"):
   # your code (log is cleared when exiting)
```

**Comment by momchil-flex** [comment]
Nice!

**Comment by momchil-flex** [comment]
I just discovered another vulnerability of the current approach. I made some modifications that made one test produce a warning, which it wasn't before. The test checks that only INFO level log is produced, however it did *not* fail with my modification. This is because the *first* thing it logs is INFO, but the next one is WARNING. However, [this code](https://github.com/flexcompute/tidy3d/blob/a9004b8c2ad1ebe028ef6c5ed7662d997c48ae65/tests/utils.py#L685-L691) returns when it sees that the first output matches the target level.

I feel like improving this may capture some other unexpected things in our tests. I'm going to move it to backlog high priority. @tylerflex since this is Shashwat's last week, should I assign you instead?

**Comment by tylerflex** [comment]
> I just discovered another vulnerability of the current approach. I made some modifications that made one test produce a warning, which it wasn't before. The test checks that only INFO level log is produced, however it did not fail with my modification. This is because the first thing it logs is INFO, but the next one is WARNING. However, [this code](https://github.com/flexcompute/tidy3d/blob/a9004b8c2ad1ebe028ef6c5ed7662d997c48ae65/tests/utils.py#L685-L691) returns when it sees that the first output matches the target level.

What would be the expected behavior for `assert_log_level(caplog, "INFO")` in your mind? it checks that the most "urgent" log level produced in `any` of the logs is `INFO`?  Could some of this be solved by just a string matching?

> I feel like improving this may capture some other unexpected things in our tests. I'm going to move it to backlog high priority. @tylerflex since this is Shashwat's last week, should I assign you instead?

Maybe we can start with what specific feature(s) do you want implemented? So far maybe
1. `assert_log_level(..., contains_str="blah")` to fail if it would otherwise pass but `contains_str` is not in the log capture?
2. look at the "maximum" log level captured and compare that to the desired level? or provide an optional argument to represent the index of the log that you want to capture?

I feel both would be pretty simple to implement, I'm just not clear on exactly what we want to do here.

**Comment by momchil-flex** [comment]
> I feel both would be pretty simple to implement, I'm just not clear on exactly what we want to do here.

I agree, I think the biggest part of the effort would be potentially updating all existing tests to the new handling.

> 1. `assert_log_level(..., contains_str="blah")` to fail if it would otherwise pass but `contains_str` is not in the log capture?
> 2. look at the "maximum" log level captured and compare that to the desired level? or provide an optional argument to represent the index of the log that you want to capture?

I feel like option 1. is much more sound. It just would require quite a lot of test updates. It solves both issues discussed here: the one I just found, as well as the one where e.g. a completely unrelated warning is raised but the test passes. Option 2. would not solve the latter. It would just be quicker to implement esp. in view of updating tests. If there's any other option that would both fit better with existing tests (not require a lot of work there) and help with both those issues, I would be OK with that too.

**Comment by tylerflex** [comment]
> I feel like option 1. is much more sound. It just would require quite a lot of test updates. It solves both issues discussed here: the one I just found, as well as the one where e.g. a completely unrelated warning is raised but the test passes. Option 2. would not solve the latter. It would just be quicker to implement esp. in view of updating tests. If there's any other option that would both fit better with existing tests (not require a lot of work there) and help with both those issues, I would be OK with that too.

We can do option 1 (or 2) quite easily and keep them both optional. Only adding it to tests in the future (or the ones you've identified already?)

**Comment by momchil-flex** [comment]
I see, yeah, ideally we should update all tests to option 1., but for starters we could just have it as an option for future tests, and at least updated existing ones to use option 2. I do feel like if e.g. an existing test is checking for `INFO` that should be the **maximum** log level there, not **any** of the log output recorded. So let's update the definition of the current `assert_log_level` to do that, and see if any tests break?

**Comment by momchil-flex** [comment]
Sorry currently it's not **any**, it's the first one, but either way I don't think that's what it should be.

**Comment by tylerflex** [comment]
Here's a [PR](https://github.com/flexcompute/tidy3d/pull/1327) to add string matching to the `assert_log_level`. This should get us started.

So the issue still remains that the code returns when it sees the first acceptable log. Momchil, maybe do you want to take a stab at defining the logic that you have in mind and committing into this PR?

**Comment by momchil-flex** [comment]
Ok, check it out: https://github.com/flexcompute/tidy3d/pull/1327/commits/d0fe26908e245451b1c4a9da821a89042ecdf74c

I also added the modification that `contains_str` is only found at the expected log level.

Three tests needed a small fix after this.

**Comment by momchil-flex** [comment]
Note that I also updated the `assert_log_level` test a bit in another commit.

**Comment by momchil-flex** [comment]
Fixed in #1223
